### PR TITLE
feat: add VoicePipelineAgent

### PR DESCRIPTION
**Pull request by nbsp**
~~depends on #99 for more work~~ transcription skipped for now

### CODE REVIEW

**Comment by nbsp** [comment]
current issues:
- [ ] sometimes buffer overflows and causes FFI panics
- [ ] interruptions do not work

File: agents/src/pipeline/agent_playout.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+          await this.#audioSource.captureFrame(frame);
+          await this.#audioSource.waitForPlayout();
+        }
+
+        // XXX(nbsp): line 161 waits instead of this. this is not the case on python agents,
+        //            but for some reason too many TTS frames can gunk up the buffer and lead to
+        //            FFI errors. this works 🤷‍♀️
+        // if (this.#audioSource.queuedDuration > 0) {
+        //   await this.#audioSource.waitForPlayout();
+        // }
```
Comment (theomonnom) [inline agents/src/pipeline/agent_playout.ts:169]: Feel free to ignore for this PR but would be good to debug what's happening. Agree that it doesn't make a ton of sense

File: examples/src/pipeline_voice_agent.ts
Comment (theomonnom) [inline examples/src/pipeline_voice_agent.ts:1]: 🔥

File: agents/src/pipeline/pipeline_agent.ts [hunk 2]
```diff
@@ -0,0 +1,917 @@
...
+        // generate an answer from the tool calls
+        extraToolsMessages.push(ChatMessage.createToolCalls(toolCallsInfo, collectedText));
+        extraToolsMessages.push(...toolCallsResults);
+
+        const chatCtx = handle.source.chatCtx.copy();
+        chatCtx.messages.push(...extraToolsMessages);
+
+        const answerLLMStream = this.llm.chat({
+          chatCtx,
+          fncCtx: i < this.#opts.maxRecursiveFncCalls - 1 ? this.fncCtx : undefined,
```
Comment (theomonnom) [inline agents/src/pipeline/pipeline_agent.ts]: We had an issue that was fixed on main. let's always give the fncCtx
```suggestion
          fncCtx: this.fncCtx,
```

File: agents/src/pipeline/pipeline_agent.ts [hunk 3]
```diff
@@ -0,0 +1,917 @@
...
+    if (!this.#started) {
+      return;
+    }
+
+    this.#room?.removeAllListeners(RoomEvent.ParticipantConnected);
+    // TODO(nbsp): await this.#deferredValidation.close()
+  }
+}
+
+async function* llmStreamToStringIterable(
```
Comment (theomonnom) [inline agents/src/pipeline/pipeline_agent.ts:813]: TIL function* is for generators in JS!

File: agents/src/pipeline/agent_output.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+// SPDX-License-Identifier: Apache-2.0
+import type { AudioFrame } from '@livekit/rtc-node';
+import { log } from '../log.js';
+import { SynthesizeStream, type TTS } from '../tts/index.js';
+import { AsyncIterableQueue, CancellablePromise, Future, gracefullyCancel } from '../utils.js';
+import type { AgentPlayout, PlayoutHandle } from './agent_playout.js';
+
+export type SpeechSource = AsyncIterable<string> | string | Promise<string>;
+
+export class SynthesisHandle {
```
Comment (theomonnom) [inline agents/src/pipeline/agent_output.ts:12]: Maybe I'm misreading the code. But it seems like classes like `SynthesisHandle`, `SpeechHandle`, `DeferredValidation` are exposed to the user. Let's keep them private if this is the case

**Review by theomonnom — APPROVED** [review]
Quite a big PR, I may have missed some stuff, but let's merge it and iterate on issues we got!

File: agents/src/pipeline/agent_output.ts (hunk 4 above)
Comment (nbsp) [reply agents/src/pipeline/agent_output.ts:12]: ah right, exported them for the other files and exported them here by accident

**Pull request merged**